radeonload.1
'\" t
.\"     Title: radeonload
.\"		Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\"		! Forked from radeontop and modified by ldsh on 27 october 2013 !
.\" Generator: DocBook XSL Stylesheets v1.77.1 <http://docbook.sf.net/>
.\"      Date: 07/11/2012
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "RADEONLOAD" "1" "07/11/2012" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
radeonload \- tool to show GPU utilization
.SH "SYNOPSIS"
.sp
\fBradeonload [\-chv] [\-b \fR\fB\fIbus\fR\fR\fB] [\-d \fR\fB\fIfile\fR\fR\fB] [\-l \fR\fB\fIlimit\fR\fR\fB] [\-t \fR\fB\fIticks\fR\fR\fB]\fR
.SH "DESCRIPTION"
.sp
RadeonLoad shows the utilization of your GPU, both in general and by blocks\&.
.sp
Supported cards are R600 and up\&.
.SH "OPTIONS"
.PP
\fB\-b, \-\-bus \fR\fB\fIbus\fR\fR
.RS 4
If you have multiple cards, use this parameter to select a specific one\&. For lspci output
\fI01:00\&.0 VGA compatible controller\fR, bus would be
\fI1\fR\&.
.RE
.PP
\fB\-c, \-\-color\fR
.RS 4
Enables colorized output\&.
.RE
.PP
\fB\-d, \-\-dump \fR\fB\fIfile\fR\fR
.RS 4
Instead of displaying the data, dump it to
\fIfile\fR\&. Use
\fI\-\fR
for stdout\&. The file is appended to, so it\(cqs safe to specify the same filename for consecutive runs\&.

Use
\fISIGTERM\fR
or
\fISIGINT\fR
for graceful termination\&.
.RE
.PP
\fB\-h, \-\-help\fR
.RS 4
Displays help\&.
.RE
.PP
\fB\-l, \-\-limit \fR\fB\fIlimit\fR\fR
.RS 4
Quit after dumping
\fIlimit\fR
lines\&. Only valid in dump mode\&. Default is to dump until terminated; use
\fISIGTERM\fR
or
\fISIGINT\fR
for graceful termination\&.
.RE
.PP
\fB\-t, \-\-ticks \fR\fB\fIticks\fR\fR
.RS 4
How many times per second to sample the data\&. The default is 120\&. Higher numbers mean more accurate output and more overhead\&.
.RE
.PP
\fB\-v, \-\-version\fR
.RS 4
Displays the program version\&.
.RE
.SH "RESOURCES"
.sp
See http://github\&.com/clbr/radeonload


radeonload.c
/*
    Copyright (C) 2012 Lauri Kasanen
    
    ! Forked from radeontop and modified  by ldsh on 27 october 2013 !

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, version 3 of the License.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "radeonload.h"
#include <getopt.h>

const void *area;

void die(const char * const why) {
	puts(why);
	exit(1);
}

static void version() {
	printf("RadeonLoad %s\n", VERSION);
	exit(1);
}

static void help(const char * const me, const unsigned int ticks) {
	printf(_("\n\tRadeonLoad for R600 and above.\n\n"
		"\tUsage: %s [-ch] [-b bus] [-d file] [-i file] [-l limit] [-t ticks]\n\n"
		"-b --bus 3		Pick card from this PCI bus\n"
		"-c --color		Enable colors\n"
		"-d --dump file		Dump data to this file, - for stdout\n"
		"-i --instant file		Dump data to this file, - for stdout\n"
		"-l --limit 3		Quit after dumping N lines, default forever\n"
		"-t --ticks 50		Samples per second (default %u)\n"
		"\n"
		"-h --help		Show this help\n"
		"-v --version		Show the version\n"),
		me, ticks);
	die("");
}

unsigned int readgrbm() {

	const void *ptr = area + 0x10;

	const unsigned int *inta = ptr;

	return *inta;
}

int main(int argc, char **argv) {

	unsigned int ticks = 120;
	unsigned char color = 0;
	unsigned char bus = 0;
	unsigned int limit = 0;
	char *dump = NULL;
	char *instant = NULL;

	// Translations
#ifdef ENABLE_NLS
	setlocale(LC_ALL, "");
	bindtextdomain("radeonload", "/usr/share/locale");
	textdomain("radeonload");
#endif

	// opts
	const struct option opts[] = {
		{"bus", 1, 0, 'b'},
		{"color", 0, 0, 'c'},
		{"dump", 1, 0, 'd'},
		{"help", 0, 0, 'h'},
		{"instant", 1, 0, 'i'},
		{"limit", 1, 0, 'l'},
		{"ticks", 1, 0, 't'},
		{"version", 0, 0, 'v'},
		{0, 0, 0, 0}
	};

	while (1) {
		int c = getopt_long(argc, argv, "b:cd:hi:lt:v", opts, NULL);
		if (c == -1) break;

		switch(c) {
			case 'h':
			case '?':
				help(argv[0], ticks);
			break;
			case 't':
				ticks = atoi(optarg);
			break;
			case 'c':
				color = 1;
			break;
			case 'b':
				bus = atoi(optarg);
			break;
			case 'v':
				version();
			break;
			case 'l':
				limit = atoi(optarg);
			break;
			case 'd':
				dump = optarg;
			case 'i':
				instant = optarg;
			break;
		}
	}

	// init
	const unsigned int pciaddr = init_pci(bus);

	const int family = getfamily(pciaddr);
	if (!family)
		puts(_("Unknown Radeon card. <= R500 won't work, new cards might."));

	const char * const cardname = family_str[family];

	initbits(family);

	// runtime
	collect(&ticks);

	if (dump)
		dumpdata(ticks, dump, limit);
	else if (instant)
		instantdata(ticks, instant, limit);
	else
		present(ticks, cardname, color);

	munmap((void *) area, MMAP_SIZE);
	return 0;
}
